elements/userlevel/todpdkdevice.cc

```diff
@@ -124,7 +124,16 @@ inline struct rte_mbuf* get_mbuf(Packet* p, bool create=true) {
 
     if (likely(DPDKDevice::is_dpdk_packet(p))) {
         mbuf = (struct rte_mbuf *) p->destructor_argument();
-        p->set_buffer_destructor(DPDKDevice::fake_free_pkt);
+        rte_pktmbuf_pkt_len(mbuf) = p->length();
```

Comment:
The length in pkt_len is the length of the data inside the packet, the buffer have a fixed size in DPDK, so packet->length() is the good one.
data_len is the size of the segment, we do not handle that for now in Click. Normally, DPDK won't give us multi-segments packets if we don't have a MTU > MBUF_DATA_SIZE. I don't have plans on managing multi-segments packets right now...

However you're right about the headroom. I added those two line to let DPDK know when the length of the packet is changed by Click, but currently if the headroom change, it does not inform DPDK. I'll add a commit to this PR. Thanks ;)